### PR TITLE
QA: fix latent bugs, harden numerics, wire telemetry, prune dead code

### PR DESCRIPTION
**Pull request by randyap8-wq**
## Summary

A QA pass over the codebase following the large prefetch/skew/cluster/pregate refactor (PR #116) and older PRs. Each finding was verified individually before fixing. The change set is intentionally **focused and conservative** — it fixes real, demonstrable issues and wires up telemetry that was already being collected, while deliberately leaving ~80 platform-gated / feature-gated scaffolding warnings (AVX/AMX kernels, NUMA, io_uring, etc.) untouched since they are live on other targets.

`cargo test`: **466 unit + 4 concurrency-stress pass.** Net unique `--all-targets` warnings **89 → 81**, none newly introduced.

## Bugs / correctness

- **`model.rs` — V projection bias loaded with the wrong size (latent data-loss bug).** The `v_proj.bias` tensor was loaded against the `kv_dim` (`num_kv_heads * head_dim`) expectation, but its real length is `v_proj_dim` (`num_kv_heads * v_head_dim`). For an **asymmetric-V** checkpoint (`v_head_dim != head_dim`, e.g. MiMo-V2-Flash) with `attention_bias = true`, the size-checked `find_f32` helper **silently dropped the bias**, degrading output with no error. Fixed to use the already-computed `v_proj_dim`. **Added a regression test** that fails without the fix.

## Numerical hardening

- **`transformer.rs` / `mla.rs` `softmax_inplace` — NaN on a fully-masked row.** A row that is entirely `-inf` (e.g. fully masked attention) computed `(-inf) - (-inf) = NaN`. Both now fall back to a uniform distribution when the row max is non-finite. **Tests added for both.**
- **`model.rs` `validate` — zero-width RoPE.** A `partial_rotary_factor` that rounds the rotary width down to `0` would silently disable rotary embeddings. Now rejected at validation time. **Test added.**
- **`pregate.rs`** — guard the consecutive-layer link against the `u32::MAX` sentinel wrapping around to `0`.
- **`io_provider.rs` `packed_offset`** — return `u64::MAX` (forcing a loud read error) instead of `0` (which would silently alias expert 0's bytes) for an id missing from the packed manifest.

## Telemetry (collected but never surfaced)

- Surface **`singleflight_followers`** (redundant SSD reads avoided via request coalescing) through `EngineReport` and the run summary.
- Surface the **prefetch-drop / gpu-cpu-fallback / speculator-mismatch / expert-read-failure** health counters as a conditional summary line.

Both summary additions are **gated** so the legacy summary shape is preserved on clean runs and only appears when there is something to report.

## Dead code

- Remove redundant `EngineReport::{total_io_wait_us, total_cycle_us}` (never read; derivable from the per-token average + token count).
- Remove orphaned `ExpertResident::{raw, header}` (zero callers, referenced a non-existent dump tool).
- Remove an unused `parking_lot::Mutex` import and an unnecessary `mut`.
- Remove 5 unused `ExpertCache` test-module imports.

## Notes / scope

- Platform- and feature-gated dead-code warnings (AVX/AMX kernels, NUMA Linux fns, io_uring) were **left in place** — they are live on x86_64/Linux and removing them would break those targets.
- Several flagged items were investigated and **rejected as false positives** (e.g. speculator training helpers that are reached via the prod train queue, `ExpertCache` pin/unpin helpers used in prod or `#[allow]`-ed test-only).

🤖 Generated with QA audit pass.

## Release Notes

* **Bug Fixes**
  * Fixed softmax numeric stability when processing fully-masked rows to prevent NaN propagation
  * Corrected attention bias tensor size calculation during model loading for asymmetric configurations
  * Improved error handling for missing expert data with explicit failure signaling

* **New Features**
  * Added SSD read deduplication telemetry to engine reporting
  * Enhanced engine health diagnostics with aggregated degradation signal monitoring

* **Validation**
  * Stricter model configuration validation to prevent silent feature disabling

* **Tests**
  * Expanded test coverage for numeric stability edge cases and model loading scenarios

### CODE REVIEW

File: rust-engine/src/mla.rs [hunk 1]
```diff
@@ -52,6 +52,13 @@ fn softmax_inplace(scores: &mut [f32]) {
         return;
     }
     let max = scores.iter().copied().fold(f32::NEG_INFINITY, f32::max);
+    // Fully-masked row (every score `-inf`): fall back to a uniform
+    // distribution rather than propagating `NaN`s from `(-inf) - (-inf)`.
+    if !max.is_finite() {
+        let uniform = 1.0 / scores.len() as f32;
+        scores.iter_mut().for_each(|s| *s = uniform);
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (randyap8-wq) [reply rust-engine/src/mla.rs]: Good catch, fixed in cba558c. The max scan now sets a `saw_nan` flag and falls back to uniform on `saw_nan || !max.is_finite()`, covered by a new `softmax_mixed_nan_is_uniform` regression test. I intentionally kept the simple uniform fallback for the `+inf` case rather than the proportional-mass distribution: routing `+inf` through the same fallback already guarantees a NaN-free output without the extra branch.

File: rust-engine/src/transformer.rs [hunk 2]
```diff
@@ -1446,6 +1446,14 @@ pub fn softmax_inplace(v: &mut [f32]) {
...
         }
     }
+    // Fully-masked or otherwise non-finite input (every logit `-inf`, or a
+    // stray `NaN`): emit a uniform distribution instead of letting the
+    // `x - max` subtraction produce `NaN`s that then propagate downstream.
+    if !max.is_finite() {
+        let uniform = 1.0 / v.len() as f32;
+        v.iter_mut().for_each(|x| *x = uniform);
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (randyap8-wq) [reply rust-engine/src/transformer.rs]: Accepted the core fix in cba558c: a `saw_nan` flag now triggers the uniform fallback even when `max` stays finite, covered by a new `softmax_mixed_nan_is_uniform` test. I declined the `+inf` proportional-distribution piece on purpose; `+inf` routes through the uniform fallback, which is enough to keep the output NaN-free without the added branch.

File: rust-engine/src/transformer.rs [hunk 3]
```diff
@@ -1446,6 +1446,14 @@ pub fn softmax_inplace(v: &mut [f32]) {
             max = x;
         }
     }
+    // Fully-masked or otherwise non-finite input (every logit `-inf`, or a
+    // stray `NaN`): emit a uniform distribution instead of letting the
+    // `x - max` subtraction produce `NaN`s that then propagate downstream.
+    if !max.is_finite() {
+        let uniform = 1.0 / v.len() as f32;
```
[the remark this answers is not in the document]
Comment (randyap8-wq) [reply rust-engine/src/transformer.rs]: Done in cba558c. The scan now seeds `max` from `NEG_INFINITY` and sets a `saw_nan` flag, so a NaN anywhere triggers the uniform fallback and the result is no longer index-order dependent (previously a NaN at `v[0]` behaved differently from one later in the row). Masked `-inf` rows with at least one finite logit still take the normal path. Added a mixed-NaN regression test.

File: rust-engine/src/mla.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (randyap8-wq) [reply rust-engine/src/mla.rs]: Done in cba558c. I track `saw_nan` while computing `max` and trigger the uniform fallback on `saw_nan || !max.is_finite()`, so a stray NaN no longer poisons the row while ordinary masked `-inf` entries still take the normal path. Added a `softmax_mixed_nan_is_uniform` regression test.
